### PR TITLE
add a manual test for hr-time origin

### LINKED ISSUE #32: Add prompt to unload test
In reading the L2 spec, I believe we are missing a [prompt to unload](https://w3c.github.io/hr-time/#h-time-origin-1) test. We should verify that we have UAs shipping this piece of the spec.

### PR DESCRIPTION
**Pull request by siusin**
This test aims to fix [hr-time/issues/32](https://github.com/w3c/hr-time/issues/32).

This [idea](https://github.com/w3c/hr-time/issues/32#issuecomment-298778029) came from @toddreifsteck and @zcorpan (to use window.opener for same origin communication).

### CODE REVIEW

**Comment by igrigorik** [comment]
@siusin is this staged on w3c-test.org?

**Comment by siusin** [comment]
@igrigorik it should be done automatically... perhaps in a few minutes.

**Comment by siusin** [comment]
@igrigorik @toddreifsteck 
mirror in the w3c server: [https://www.w3.org/2017/04/webperf-test/unload_manual.html](https://www.w3.org/2017/04/webperf-test/unload_manual.html )

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242909320)

# Firefox (nightly) #
Testing web-platform-tests at revision 49e18ca34e247b954302fce2800e1fffa142a696
Using browser at version BuildID 20170612100241; SourceStamp 27cad9749cddf68e11fdd4e5d73dad84a8f8cf23
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6241/hr-time/unload_manual.html">/hr-time/unload_manual.html</a></summary>


|                                                               Subtest                                                                | Results | Messages |
|--------------------------------------------------------------------------------------------------------------------------------------|---------|----------|
|                                                                                                                                      | TIMEOUT |          |
| `Window time origin is almost equal (<10ms) to the time of the user confirming the navigation if a confirmation dialog is displayed` | NOTRUN  |          |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242909321)

# Sauce (safari) #
Testing web-platform-tests at revision 49e18ca34e247b954302fce2800e1fffa142a696
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6241/hr-time/unload_manual.html">/hr-time/unload_manual.html</a></summary>


|                                                               Subtest                                                                | Results | Messages |
|--------------------------------------------------------------------------------------------------------------------------------------|---------|----------|
|                                                                                                                                      | TIMEOUT |          |
| `Window time origin is almost equal (<10ms) to the time of the user confirming the navigation if a confirmation dialog is displayed` | NOTRUN  |          |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242909333)

# Chrome (unstable) #

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242909334)

# Sauce (MicrosoftEdge) #

**Comment by zcorpan** [comment]
http://w3c-test.org/submissions/6241/hr-time/unload_manual.html

**Comment by zcorpan** [comment]
@siusin can you please check the checkbox "Allow edits from maintainers" in the sidebar, and comment when you have done so? I have some fixes I want to push to this branch.

**Comment by siusin** [comment]
@zcorpan Done. Thanks & look forward to your review.

File: hr-time/unload_manual.html
Comment (spanicker) [inline hr-time/unload_manual.html]: Confused by this - this is a manual test right? 
It's not possible for a human to confirm the dialog within 10ms?

Comment (siusin) [inline hr-time/unload_manual.html]: @spanicker sorry, this is should be 10s, the [default timeout](http://web-platform-tests.org/writing-tests/testharness-api.html#harness-timeout) of testharness.

File: hr-time/unload_manual.html
Comment (spanicker) [inline hr-time/unload_manual.html]: This is still using 10ms?
The test is defined as "time origin is almost equal (<10ms) to the time of the user confirming the navigation" -- this is not really viable

File: hr-time/resources/unload_child.html
Comment (igrigorik) [inline hr-time/resources/unload_child.html]: A little confused by this.. 

- origin == "child's time origin mapped into parent's origin". right? Should we spell this out in the test?
- "unload time was": "child's unload time mapped into parent's origin". right? Same as above.

---

Running the test, I see:

Index:2
origin: 3425.63
unload time was : 2709.7900000000004

This has me confused. How can the origin time be less than unload time? I think the math is not right somewhere..

File: hr-time/unload_manual.html
Comment (igrigorik) [inline hr-time/unload_manual.html]: Right, we're comparing DOMHighResTimestamps which are in milliseconds.. So that should be a much, much higher number.

**Comment by zcorpan** [comment]
OK pushed my fixes.

File: hr-time/unload-manual.html
Comment (zcorpan) [inline hr-time/unload-manual.html]: Is "within 10s" to avoid the harness timeout? I removed the timeout with `setup()` so then I suppose this doesn't need to mention any time limit.

**Comment by igrigorik** [comment]
@zcorpan the w3c-test.org link is 404'ing: http://w3c-test.org/submissions/6241/hr-time/unload_manual.html -- any chance we could get that back online?

**Comment by siusin** [comment]
Oops, it should be updated to [http://w3c-test.org/submissions/6241/hr-time/unload-manual.html](http://w3c-test.org/submissions/6241/hr-time/unload-manual.html) since we renamed the test to use -manual.html suffix.

The timeout is only there for the user interaction (confirming the reload), which shouldn't make any difference on the test result. It's better if we can remove it (thanks @zcorpan !).

I got a pass in Firefox but failed on Chrome and Safari.

File: hr-time/resources/unload_child.html
Comment (siusin) [inline hr-time/resources/unload_child.html]: This is the origin of the second navigation, and the unload time of the first navigation. Ideally,  they should be equal...

**Comment by zcorpan** [comment]
I did a second pass to simplify the test quite a bit. I removed the `innerHTML` debugging stuff; the actual values are printed in the summary table, and devtools can be used for debugging the test (and then it helps to have less code).

**Review by zcorpan — DISMISSED** [review]

**Comment by zcorpan** [comment]
In EdgeHTML 15 I get three confirmation dialogs with the test; first the expected reload prompt, then a dialog to confirm `window.close()`, then another unload prompt (here `onbeforeunload` has been set to `null` so this prompt shouldn't appear). The test also fails.

Can you file bugs on browsers that fail the test?

File: hr-time/resources/unload_child.html
Comment (igrigorik) [inline hr-time/resources/unload_child.html]: Isn't this still testing <10ms delta? We want it to be 10s -- right?

**Comment by toddreifsteck** [comment]
Hosted version doesn't appear to be update to date

**Comment by toddreifsteck** [comment]
up to date

File: hr-time/resources/unload_child.html
Comment (zcorpan) [inline hr-time/resources/unload_child.html]: No. It should be close to 0, as I understand it. 10s was unrelated (default testharness.js timeout, now disabled).

**Comment by zcorpan** [comment]
The one on w3c-test.org is. Otherwise you can check out this branch and test locally.

**Comment by siusin** [comment]
@toddreifsteck @zcorpan 
The test is now available as [http://w3c-test.org/submissions/6241/hr-time/unload-manual.html](http://w3c-test.org/submissions/6241/hr-time/unload-manual.html) on w3c-test.org.

Does the link work for you?

**Comment by igrigorik** [comment]
> [hr-time] the time of the user confirming the navigation if a confirmation dialog is displayed during the prompt to unload algorithm;

Hmm, testing in Chrome.. @spanicker it seems that we may be capturing when the dialogue is displayed, not when it's accepted.

**Comment by siusin** [comment]
@spanicker pointed out that onbeforeunload is not the time when the user confirms the dialog, we might need to capture the time of the click... 

@toddreifsteck suggested we can take out the time of navigation.upload()... hopefully he will submit another fix later.

**Comment by igrigorik** [comment]
@siusin @toddreifsteck navigation.upload? Any updates on this test?

**Comment by toddreifsteck** [comment]
The code in this branch is very close. I've scheduled an hour on 7/12 to finish this out. I'll report back after I'm done.

**Comment by igrigorik** [comment]
@toddreifsteck any luck or updates on this one?

**Comment by divmain** [comment]
@igrigorik I'm looking into this on @toddreifsteck's behalf.

**Comment by igrigorik** [comment]
@divmain awesome, let me know if and how I can help to land this.

**Comment by igrigorik** [comment]
@divmain any luck on your end? :)

**Comment by plehegar** [comment]
w3c-test:mirror

**Comment by igrigorik** [comment]
*bump* 

@divmain @toddreifsteck any updates on this one? :)

**Comment by divmain** [comment]
Sorry it has taken me so long to get to this.  I did a cursory review and discovered a potential ambiguity in the spec, and after that it just got away from me.  Here's what I've got...

### Time Origin

The [time origin](https://w3c.github.io/hr-time/#time-origin-0), where the global object is a `Window`, is the pertinent part of the spec.  That section is as follows, with bullet points labeled for easier reference.

-----

The time origin must be equal to:

<ol type="i">
  <li>the time when the browsing context is first created if there is no previous document;</li>
  <li>otherwise, the time of the user confirming the navigation if a confirmation dialog is displayed during the prompt to unload algorithm;</li>
  <li>otherwise, the time of starting the navigation responsible for loading the Window object's newest Document object.</li>
</ol>

-----

### Life-cycle Events

The following is a diagram of the relevant DOM life-cycle events:

performance-measurements

With that as context, I'll go through the various scenarios.

-----

### Scenarios

**No previous document:**

This addresses condition `i`, and is relatively straightforward and unambiguous.  Should `now()` be called at point `B`, `C`, and `E` in the diagram above, the return values would be `A ➡ B`, `A ➡ C`, and `A ➡ E`, respectively.

**Previous document and prompt displayed:**

This addresses condition `ii`, where condition `i` is untrue, i.e.

- there _is_ a previous document,
- a truthy value was returned in the handler for event `C`,
- a prompt was displayed, and
- the user opted to allow the navigation (`D`).

Should `now()` be called at point `G`, the return value would be `D ➡ G`.

**Previous document and prompt _not_ displayed:**

This addresses condition `iii`, where conditions `i` and `ii` are untrue, i.e.

- there _is_ a previous document,
- a handler for event `C` either does not exist or returned a falsey value, and
- _no_ prompt was displayed to the user (`D`).

Should `now()` be called at point `G`, the return value would be `F ➡ G`.

**_Two_ previous documents and prompt displayed:**

This is where the ambiguity comes in.  Consider a variation of the diagram above (irrelevant life-cycle events are elided for brevity):

performance-measurements-b

This scenario focuses on `Document 2` and event `EE` above.  In this scenario, the user has navigated at least once before the navigation in question (`Document 1`/`CC`) and wants to navigate again:

- a user clicks a link in `Document 2`,
- a truthy value is returned from the `onbeforeunload` handler,
- the user opts to allow navigation (`DD`), and
- the window navigates to `Document 3` (`FF`).

Should `now()` be called at point `EE`, **the return value could be one of two values**, depending on how you read the spec: `BB ➡ EE` or `DD ➡ EE`.

Consider the condition for `i`: `... if there is no previous document`.  At `EE`, this condition is false; there _is_ a previous document.

That brings us to `ii`: `the time of the user confirming the navigation if a confirmation dialog is displayed during the prompt to unload algorithm`.

By a strict reading of the spec, this rule is active at `EE`: a confirmation dialog was displayed during the prompt to unload.  This means the return value of a call to `now()` at `EE` should be `DD ➡ EE`.  However, that does not appear to be the implemented behavior in any browser I checked.  To test this, insert the following into the console of your browser:

```javascript
window.onbeforeunload = function () {
  console.log(performance.now());
  return true;
};
window.onunload = function () {
  console.log(performance.now());
};
```

The second value printed to the console will be larger than the first.

### Testing

My guess is that the spec should be updated to reflect the implemented behavior, but I will leave that to the working group to determine.  Should that be the case, the following is some pseudo-code for what a good test might look like.

**=== Setup ===**

- The following handlers are specified for `child-a.html`'s events:
    + on `window`
        * `beforeunload`
            - `window.opener.postMessage` with `event: "beforeunload"` and `value: performance.now()`
            - `return true`
        * `unload`
            - `window.opener.postMessage` with `event: "unload"` and `value: performance.now()`
            - add a blocking `while` loop that continues looping until two full seconds have elapsed, delaying navigation
- The following handlers are specified `child-b.html`'s events:
    + on `window`
        * `load`
            - `window.opener.postMessage` with `event: "load-b"` and `value: performance.now()`
- The following handlers are specified for `parent.html`'s events:
    + on `button`
        * `click`
            - `./resources/child-a.html` is opened in a new window/tab
            - a reference to this child window is saved for later
    + on `window`
        * `message`
            - if `event` is `beforeunload`, store that value
            - if `event` is `unload`, store that value
            - if `event` is `load-b`
                + store the value
                + close the child window (via the previously saved reference)
                + run assertions

**=== Test Run ===**

- User opens `parent.html`.
- User clicks button.
- A new window is opened for `child-a.html`.
- User clicks a link directing them to `child-b.html`.
- User is prompted to specify whether they'd like to leave the page, and they agree.
- Child window navigates to `child-b.html`, which is quickly closed by the parent window.

**=== Assertions ===**

- the `value` of `unload` is larger than the `value` of `beforeunload`
- the `value` of `load-b` is larger than `2000` (the `unload` handler should have consumed at least 2 seconds)

**Comment by divmain** [comment]
As for this PR, the child's time origin is being compared to the opener's (parent's) origin.  This seems to miss the mark, unless I'm missing something: the opener is not the previous document referenced in the spec, and therefore should have no bearing on any measurements done in the child.

Instead, the child should have a second navigation event and the measurements should be compared as indicated in my previous comment.

**Comment by igrigorik** [comment]
@divmain thanks for the detailed writeup! I'll have to noodle on this one a bit.. 

I've added this to our tentative agenda for WG call this coming Monday (Sept 25th @ 10AM PST) -- any chance you could join to discuss?

**Comment by divmain** [comment]
Yep, I plan to be on the call.

**Comment by plehegar** [comment]
Discussed at https://www.w3.org/2017/09/25-webperf-minutes.html#item04

**Comment by foolip** [comment]
@siusin, should this be closed given that https://github.com/web-platform-tests/wpt/pull/8488 was merged, or is there still something to be merged here?

**Comment by siusin** [comment]
Closed. Thanks @foolip for the ping.

**Pull request closed without merging**
